component/loki/source/kubernetes/kubetail/tailer.go

```diff
@@ -167,9 +169,51 @@ func (t *tailer) tail(ctx context.Context, handler loki.EntryHandler) error {
 	if err != nil {
 		return err
 	}
+
+	// Create a new rolling average calculator to determine the average delta
+	// time between log entries.
+	//
+	// Here, we track the most recent 10,000 delta times to compute a fairly
+	// accurate average. If there are less than 100 deltas stored, the average
+	// time defaults to 1h.
+	//
+	// The computed average will never be less than the minimum of 2s.
+	calc := newRollingAverageCalculator(10000, 100, 2*time.Second, maxTailerLifetime)
+
 	go func() {
-		<-ctx.Done()
-		_ = stream.Close()
+		rolledFileTicker := time.NewTicker(1 * time.Second)
+		defer func() {
+			rolledFileTicker.Stop()
+			_ = stream.Close()
+		}()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-rolledFileTicker.C:
+				// Versions of Kubernetes which do not contain
+				// kubernetes/kubernetes#115702 will fail to detect rolled log files
+				// and stop sending logs to us.
+				//
+				// To work around this, we use a rolling average to determine how
+				// frequent we usually expect to see entries. If 3x the normal delta has
+				// elapsed, we'll restart the tailer.
+				//
+				// False positives here are acceptable, but false negatives mean that
+				// we'll have a larger spike of missing logs until we detect a rolled
+				// file.
+				avg := calc.GetAverage()
+				last := calc.GetLast()
+				if last.IsZero() {
+					continue
+				}
+				s := time.Since(last)
+				if s > avg*3 {
+					level.Info(t.log).Log("msg", "have not seen a log line in 3x average time between lines, closing and re-opening tailer", "rolling_average", avg, "time_since_last", s)
+					return
```

Comment:
Just to make sure I get this correctly, this will refresh the tailer, because the deferred function will call stream.Close(), causing line 263 below to return and the `retailBackoff` triggering another tailer to start, right?

Reply:
Yes 👍